### DISCUSSION TITLE
Access array of page components to build dynamic features

### DISCUSSION THREAD

**Discussion by thebergamo** [vercel/next.js#38893] [Ideas]
### Describe the feature you'd like to request

When using react-router or other routers that you can for example read your page components and centralize this configurations in one file it makes easier to for instance create a side menu dynamically.

Next.js brings us a real nice way of avoiding this central place to manage routes using a file system based routing engine, but this without a way for example to access this list of routes or even a way to access some special properties from our page components dynamically makes it a bit harder to take further advantage of such router.

At this moment, what exactly I'm trying to achieve? I would like to build a side-menu that add items dynamically based on my file system routes. (Without keeping a static file) 

### Describe the solution you'd like

I would like to be able to for example, have a list of all routes my app have and also some custom properties being available.

For example:
```tsx
// /pages/user/index.tsx

const UserListPage = () => <h1>Yay!</h1>

export default UserListPage;

export const sideMenu = true;
export const sideMenuLabel = 'Users';
export const SideMenuIcon = 'user'; // Or <UserIcon />
```

then on my for example `/components/SideMenu.tsx`

```tsx
export const SideMenu = () => {
    const router = useRouter();
    const pages = router.getAllPages();
    const menuItems = pages.filter(({ sideMenu }) => sideMenu)

    return (<>
         {menuItems.map(({ path, sideMenuLabel, SideMenuIcon }) => (<Link to={path} key={path}>
               <div className={path === router.asPath && 'active'}>
                      <SideMenuIcon />
                      <h3>{sideMenuLabel}</h3>
               </div>
            </Link>
    </>);

}
```

### Describe alternatives you've considered

An alternative to this approach would be maintaining a static configuration file for this, but then everytime we add a new route we would need to update this file, while the dynamic routing from Next.js could help.

Another alternative that I tried was to read all pages and import the component to read a property from it as per #3351 but no success.

**Comment by Babbili** [comment]
@thebergamo You can achieve that in Next Js by creating a custome [Layout component](https://nextjs.org/docs/basic-features/layouts) which will be shared across all your app pages, include the `<SideMenu />` component in this Layout
```tsx
//     components/Layout.tsx

import React from 'react'
import SideMenu from './SideMenu.tsx'

const Layout: React.FC<any> = ({ children }) => {

 return(
    <>
        <SideMenu />    // include the side menu in you app layout
        <main>{ children }</main>
    </>
 )
}
export default Layout
```
create a [Custom App](https://nextjs.org/docs/advanced-features/custom-app) and wrap your application with the layout. Since the `<Layout />` component is re-used when changing pages, its component state will be preserved
```tsx
//   pages/_app.tsx

import '../styles/globals.css'
import type { AppProps } from 'next/app'
import Layout from '../components/Layout'

function MyApp({ Component, pageProps }: AppProps) {
  return(
    <Layout>
      <Component {...pageProps} />
    </Layout>
  )
}

export default MyApp
```

**Reply by thebergamo** [reply]
That's not really the point, yes I can create this component, but how I can load all possible routes into it?

How I can for example know without setup a static json or any other way get the routes that are available inside my pages folder?

I think I could write a script to load it in build time, but I believe best would be NextJS providing us a way to extract those dynamic info on run time.

That could also provide us a nice way to load new features inside an app without restarting it.
